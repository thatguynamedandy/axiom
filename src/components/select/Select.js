import React, { Component, PropTypes, cloneElement } from 'react';
import {
  TextInputIcon,
  Context,
  ContextMenu,
  Dropdown,
  DropdownButton,
  DropdownMenu,
} from 'bw-axiom';
import SelectInput from './SelectInput';

export default class Select extends Component {
  static propTypes = {
    children: PropTypes.node,
  };

  constructor(props) {
    super(props);

    const index = Math.max(props.children.findIndex((child) => child.props.selected), 0);

    this.state = {
      focusedIndex: index,
      value: props.children[index].props.value,
    };
  }

  handleMouseEnter(index) {
    this.setState({ focusedIndex: index });
  }

  handleSelect() {
    this.setState(({ focusedIndex }, { children }) =>
      ({ value: children[focusedIndex].props.value }));
  }

  focusNext() {
    this.setState(({ focusedIndex }, { children }) =>
      ({ focusedIndex: (focusedIndex === children.length - 1 ) ? 0 : focusedIndex + 1 }));
  }

  focusPrev() {
    this.setState(({ focusedIndex }, { children }) =>
      ({ focusedIndex: (focusedIndex === 0) ? children.length - 1 : focusedIndex - 1 }));
  }

  filter(str) {
    const regex = new RegExp(`^${ str }`);
    this.setState(({ focusedIndex }, { children }) => {
      const match = children.findIndex(({ props }) => regex.test(props.value));
      return { focusedIndex: (match >= 0) ? match : focusedIndex };
    });
  }

  render() {
    const { children, ...rest } = this.props;
    const { focusedIndex, value } = this.state;

    const mappedChildren = children.map((child, index) =>
      cloneElement(child, {
        onClick: ::this.handleSelect,
        onMouseEnter: this.handleMouseEnter.bind(this, index),
        selected: child.props.value === value,
        focused: index === focusedIndex,
      }));

    return (
      <Dropdown>
        <DropdownButton>
          <SelectInput
              filter={ ::this.filter }
              focusNext={ ::this.focusNext }
              focusPrev={ ::this.focusPrev }
              handleSelect={ ::this.handleSelect }
              value={ value }
              { ...rest }>
            <TextInputIcon name="chevron-down" />
          </SelectInput>
        </DropdownButton>

        <DropdownMenu>
          <Context position="bottom">
            <ContextMenu>
              { mappedChildren }
            </ContextMenu>
          </Context>
        </DropdownMenu>
      </Dropdown>
    );
  }
}
