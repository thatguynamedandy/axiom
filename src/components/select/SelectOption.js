import PropTypes from 'prop-types';
import React, { Component } from 'react';
import DropdownMenuItem from '../dropdown/DropdownMenuItem';

export default class SelectOption extends Component {
  static propTypes = {
    children: PropTypes.node,
    onClick: PropTypes.func,
    value: PropTypes.any.isRequired,
  };

  static contextTypes = {
    selectedOptionValue: PropTypes.any,
    handleSelectOption: PropTypes.func.isRequired,
  };

  constructor(props) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick(event) {
    const { onClick, value } = this.props;

    this.context.handleSelectOption(value);
    if (onClick) onClick(event);
  }

  render() {
    const { children, value, ...props } = this.props;
    const { selectedOptionValue } = this.context;

    const selected =
      (Array.isArray(selectedOptionValue) && selectedOptionValue.indexOf(value) !== -1)
      || value === selectedOptionValue;

    return (
      <DropdownMenuItem
          { ...props }
          onClick={ this.handleClick }
          selected={ selected }>
        { children }
      </DropdownMenuItem>
    );
  }
}
