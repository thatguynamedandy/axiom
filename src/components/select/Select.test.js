import React from 'react';
import renderer from 'react-test-renderer';
import Select from './Select';
import SelectOption from './SelectOption';

const getComponent = (props = {}) =>
  renderer.create(
    <Select { ...props }>
      <SelectOption value="1" />
      <SelectOption value="2" />
      <SelectOption value="3" />
    </Select>
  );

describe('Select', () => {
  it('renders with defaultProps', () => {
    const component = getComponent();
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with a selected value', () => {
    const component = getComponent({ selectedValue: '2' });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
