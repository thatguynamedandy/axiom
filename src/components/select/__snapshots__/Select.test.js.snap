// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders with a selected value 1`] = `
<label
  className="ax-input__group ax-space--x4"
>
  <div
    className="ax-input__button-container"
  >
    <div
      className="ax-input__icon-container ax-input__icon-container--medium"
    >
      <span
        className="ax-input__icon ax-input__icon--align-right"
      >
        <svg
          className="ax-icon ax-icon--chevron-down"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<path d=\\"M4 5L2.5 6.5 8 12l5.5-5.5L12 5 8 9z\\" fill-rule=\\"nonzero\\" fill=\\"currentColor\\"/>",
            }
          }
          style={
            Object {
              "height": "1rem",
              "width": "1rem",
            }
          }
          viewBox="0 0 16 16"
        />
      </span>
      <input
        className="ax-input"
        disabled={undefined}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
</label>
`;

exports[`Select renders with defaultProps 1`] = `
<label
  className="ax-input__group ax-space--x4"
>
  <div
    className="ax-input__button-container"
  >
    <div
      className="ax-input__icon-container ax-input__icon-container--medium"
    >
      <span
        className="ax-input__icon ax-input__icon--align-right"
      >
        <svg
          className="ax-icon ax-icon--chevron-down"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<path d=\\"M4 5L2.5 6.5 8 12l5.5-5.5L12 5 8 9z\\" fill-rule=\\"nonzero\\" fill=\\"currentColor\\"/>",
            }
          }
          style={
            Object {
              "height": "1rem",
              "width": "1rem",
            }
          }
          viewBox="0 0 16 16"
        />
      </span>
      <input
        className="ax-input"
        disabled={undefined}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
</label>
`;
